db/src/timelines.rs

```diff
@@ -0,0 +1,442 @@
+
+use rusqlite;
+
+use errors::{
+    DbErrorKind,
+    Result,
+};
+
+use mentat_core::{
+    Entid,
+    Schema,
+    TypedValue,
+    KnownEntid,
+};
+
+use edn::{
+    InternSet,
+};
+
+use edn::entities::OpType;
+
+use db;
+use db::{
+    TypedSQLValue,
+};
+
+use tx::{
+    transact_terms_with_action,
+    TransactorAction,
+};
+
+use types::{
+    PartitionMap,
+};
+
+use internal_types::{
+    Term,
+    TermWithoutTempIds,
+};
+
+use watcher::{
+    NullWatcher,
+};
+
+pub static MAIN_TIMELINE: Entid = 0;
+
+fn ensure_on_tail_of(conn: &rusqlite::Connection, sorted_tx_ids: &[Entid], timeline: Entid) -> Result<()> {
+    let mut stmt = conn.prepare("SELECT tx, timeline FROM timelined_transactions WHERE tx >= ? AND timeline = ?")?;
+    let mut rows = stmt.query_and_then(&[&sorted_tx_ids[0], &timeline], |row: &rusqlite::Row| -> Result<(Entid, Entid)>{
+        Ok((row.get_checked(0)?, row.get_checked(1)?))
+    })?;
+
+    // Ensure that tx_ids are a consistent block at the tail end of the transactions table.
+    // TODO do this in SQL? e.g. SELECT tx FROM timelined_transactions WHERE tx >= ? AND tx NOT IN [tx_ids]...
+    let timeline = match rows.next() {
+        Some(t) => {
+            let t = t?;
+            if !sorted_tx_ids.contains(&t.0) {
+                println!("1) got tx: {}, list of txs: {:?}", t.0, sorted_tx_ids);
+                bail!(DbErrorKind::TimelinesNotOnTail);
+            }
+            t.1
+        },
+        None => bail!(DbErrorKind::TimelinesInvalidTransactionIds)
+    };
+
+    while let Some(t) = rows.next() {
+        let t = t?;
+        if !sorted_tx_ids.contains(&t.0) {
+            println!("2) got tx: {}, list of txs: {:?}", t.0, sorted_tx_ids);
+            bail!(DbErrorKind::TimelinesNotOnTail);
+        }
+        if t.1 != timeline {
+            bail!(DbErrorKind::TimelinesMixed);
+        }
+    }
+
+    Ok(())
+}
+
+fn move_transactions_to(conn: &rusqlite::Connection, tx_ids: &[Entid], new_timeline: Entid) -> Result<()> {
+    // Move specified transactions over to a specified timeline.
+    conn.execute(&format!(
+        "UPDATE timelined_transactions SET timeline = {} WHERE tx IN ({})",
+            new_timeline,
+            ::repeat_values(tx_ids.len(), 1)
+        ), &(tx_ids.iter().map(|x| x as &rusqlite::types::ToSql).collect::<Vec<_>>())
+    )?;
+    Ok(())
+}
+
+/// Get terms for tx_id, reversing them in meaning (swap add & retract).
+fn reversed_terms_for(conn: &rusqlite::Connection, tx_id: Entid) -> Result<Vec<TermWithoutTempIds>> {
+    let mut stmt = conn.prepare("SELECT e, a, v, value_type_tag, tx, added FROM timelined_transactions WHERE tx = ? AND timeline = ? ORDER BY tx DESC")?;
+    let mut rows = stmt.query_and_then(&[&tx_id, &MAIN_TIMELINE], |row| -> Result<TermWithoutTempIds> {
+        let op = match row.get_checked(5)? {
+            true => OpType::Retract,
+            false => OpType::Add
+        };
+        Ok(Term::AddOrRetract(
+            op,

```

Comment:
This is where we could use that row helper, although I see you're mapping as well.